src/fs/filesystem.ts

```diff
@@ -157,6 +161,7 @@ export class FileSystem {
   }
 
   async sync(): Promise<CID> {
+    // waterfall? Why not in parallel?
     this.root = await asyncWaterfall(this.root, [
```

Comment:
`addChild` returns an the updated tree. I don't think that we'd want to run these exact operations in parallel because they could get out of sync.
But I agree about the concern of a bottleneck. Instead we could either:
- `put` each child, and then manually update the links in the `Tree` all at once
- add a tree method for adding multiple children in parallel (which would essentially do the above^ but in the `Tree` implementation instead of here)

Reply:
> because they could get out of sync

How? Why would there be dependencies between these?

Reply:
Let's say you add 2 files: `deeply/nested/file1` and `deeply/nested/file2` at the same time

You tell the root node to add both files, for each it recurses down to `deeply/nested` & adds the file, then recursively updates the nodes on the way back. You're going to end up splitting the tree because `file1` & `file2` aren't getting added to the same `nested` object, they're being added to 2 separate objects that both represent `nested`. 

You either need to update `nested` for both `file1` and `file2` in the same operation. Or you need to make sure that the operation that is adding `file2` to `nested` is aware that `nested` has been updated to include `file1`.

This could be fixed in the current implementation by adding an `addChildren` method that takes multiple children for adding children in parallel. Or by making `Tree`s more stateful so that each operation will be aware of the other one happening, altho we might run into some weird race conditions with something like this...

I think that the plan to serialize the FS to a list before adding it to the network work going to make this a lot simpler

Reply:
> at the same time

How would you even do that? Don't you need to call some kind of "add" on it twice?

I'm still unclear why you'd need this to be stateful at all. Each subtree is independent, even in your description above. You recurse through them recursively with a base case. They then build a new nested object. We should be able to blow away the entire structure and build a new one fresh if we need to, and get the same result, no?

Reply:
> I think that the plan to serialize the FS to a list before adding it to the network work going to make this a lot simpler

I'm very heavily leaning to not using a list for the public side, only the private one

Reply:
> it recurses down to deeply/nested & adds the file

Actually, is this happening with structural recursion, or message passing? 👀

Reply:
> How would you even do that?
by "at the same time", I meant "in parallel".

> I'm still unclear why you'd need this to be stateful at all. Each subtree is independent, even in your description above. 

The folder `/deeply/nested` is being touched by two separate operations though. So if they're running in parallel, they can't be treated independently since they're both altering the same directory

> You recurse through them recursively with a base case. They then build a new nested object. We should be able to blow away the entire structure and build a new one fresh if we need to, and get the same result, no?
The two `add` operations have to be aware of one another tho (for anything that is nested)
Either:
- `file1` and `file2` need to be added to node at the same time (this can be done sequentially in memory, but the `add` to the network would include both changes at the same time)
- or whatever function is adding `file2` to `deeply/nested` must be aware that `file1` was previously added to it.
Otherwise the tree will split and one function will think that `deeply/nested` is a dir that contains one file: `file1` and the other function will think that its a dir that contains one file: `file2`

Am I understanding your point correctly? It seems like what we're each saying isn't quite clicking :sweat_smile:

Reply:
> I'm very heavily leaning to not using a list for the public side, only the private one. We can almost certainly get that simplicity by making the structure less stateful / more independent
ok cool :+1: I'm interested to hear why the change in the thought, cause last I remember, we were leaning towards keeping the implementations the same. But maybe this would be better to do over a different medium (Talk/Zoom)

Reply:
> Actually, is this happening with structural recursion, or message passing?
ummm not quite sure the comparison here... I'm pretty sure it's using both, but I think I'm probably misunderstanding the question :sweat_smile:

Reply:
> The folder /deeply/nested is being touched by two separate operations though. So if they're running in parallel, they can't be treated independently since they're both altering the same directory

Maybe this will clarify my thinking / possibly an incorrect assumption? 👇

1. You can't touch two nodes at once, since the memory will get lock by the runtime. The main thread is also single threaded.
2. Why this would be a problem either with immutable structural recursion or mutation. In either case, the parent should get updated. It's not possible for them to happen at the same time, and the operations commute so you should get the same result running `f . g` or `g . f`. Given this property, you can run these operations concurrently, and get the same result.